### PR TITLE
Refactor inference classes

### PR DESCRIPTION
**Pull request by fepegar**
`GridSampler` and `GridAggregator` need to be rewritten so that they're readable, bug-free and user-friendly.

### CODE REVIEW

**Comment by fepegar** [comment]
@GFabien I think the grid sampler is ready. I'll now tackle the aggregator.

**Comment by fepegar** [comment]
Something important is that I'm going to change the meaning of `patch_overlap`. It's now twice "border" in NiftyNet, so you should `8` in your code if you had `4`.

**Comment by GFabien** [comment]
> Something important is that I'm going to change the meaning of patch_overlap. It's now twice "border" in NiftyNet, so you should 8 in your code if you had 4.

I mean you use patch_overlap at the beginning and at the end so you kind of double it but it does not make sense to have the "double" value an odd value. Or maybe I didn't get how you are planning to use it

**Comment by fepegar** [comment]
Maybe some visualization will help me explain what I mean:

```
[x x x x x x x x x x x x x x x]  # image
[1 1 1 1 1]                      # patches
      [2 2 2 2 2]
            [3 3 3 3 3]
                  [4 4 4 4 4]
                    [5 5 5 5 5]
```

I think this is what happens when one sets `border` to `1`. I think if we're using the word "overlap", it makes sense to say that the overlap of this sampler is `2`.

**Comment by GFabien** [comment]
Ok I see your point, you're right.

**Comment by fepegar** [comment]
For some reason the travis build wasn't launched, by I know this is not working yet. I'll need to debug the aggregator some other day.

**Comment by fepegar** [comment]
I think some lines in `GridAggregator.crop_batch` might not be necessary, but this seems to work. @GFabien would this implementation work for you? Note the interface for the aggregator has changed, now it takes only the grid sampler.

File: torchio/data/inference/aggregator.py [hunk 1]
```diff
@@ -13,74 +13,98 @@ class GridAggregator:
...
+        num_locations = len(crop_locations)
+
+        # Do not crop patches at the border of the volume
+        border_ini = np.tile(border, (num_locations, 1))
+        border_fin = border_ini.copy()
+        mask_border_ini = indices_ini == 0
+        border_ini[mask_border_ini] = 0
+        for axis, size in enumerate(self.spatial_shape):
+            mask_border_ini = indices_fin[:, axis] == size
+            border_fin[mask_border_ini, axis] = 0
```
[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py:48]: Yes, thanks. Classic copy&paste mistake.

File: torchio/data/inference/aggregator.py [hunk 2]
```diff
@@ -13,74 +13,98 @@ class GridAggregator:
...
-        batch_shape = patches.shape
-        spatial_shape = batch_shape[2:]  # ignore batch and channels dim
-        num_dimensions = 3
-        for idx in range(num_dimensions):
-            location[:, idx] = location[:, idx] + border[idx]
-            location[:, idx + 3] = location[:, idx + 3] - border[idx]
-        cropped_shape = np.max(location[:, 3:6] - location[:, 0:3], axis=0)
+        border = np.array(overlap) // 2  # overlap is even in grid sampler
+        crop_locations = location.astype(int).copy()
+        spatial_shape = batch.shape[2:]  # ignore batch and channels dim
```
[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py:37]: Agreed.

File: torchio/data/inference/aggregator.py [hunk 3]
```diff
@@ -13,74 +13,98 @@ class GridAggregator:
...
             i_ini, j_ini, k_ini, i_fin, j_fin, k_fin = location
-            channels = len(patch)
-            for channel in range(channels):
-                self._output_tensor[channel, i_ini:i_fin, j_ini:j_fin, k_ini:k_fin] = patch[channel]
+            for channel_idx, tensor in enumerate(patch):
+                self._output_tensor[
+                    channel_idx,
+                    i_ini:i_fin,
+                    j_ini:j_fin,
+                    k_ini:k_fin] = tensor
```
[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py:99]: No, you're right.

File: torchio/data/inference/aggregator.py [hunk 4]
```diff
@@ -13,74 +13,98 @@ class GridAggregator:
...
+            mask_border_ini = indices_fin[:, axis] == size
+            border_fin[mask_border_ini, axis] = 0
+        indices_ini += border_ini
+        indices_fin -= border_fin
+
+        cropped_shape = np.max(indices_fin - indices_ini, axis=0)
         diff = spatial_shape - cropped_shape
         left = np.floor(diff / 2).astype(np.int)
         i_ini, j_ini, k_ini = left
         i_fin, j_fin, k_fin = left + cropped_shape
```
[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py:56]: I agree that was a bad solution. That's what I get for refactoring [someone else's code](https://github.com/NifTK/NiftyNet/blob/935bf4334cd00fa9f9d50f6a95ddcbfdde4031e0/niftynet/engine/windows_aggregator_base.py#L81-L144) without taking the time to properly understand it...

I've now removed that and added code to crop each patch individually, using the proper locations previously computed. Hopefully you'll agree with this solution. One issue with padding is that you need to introduce spurious intensity values in the volume, which is will be yet another hyperparameter. I suppose we could add it as a kwarg if you think it'd be worth it.

[the remark this answers is not in the document]
Comment (GFabien) [reply torchio/data/inference/aggregator.py]: Nice, I'm good with your new version! I agree that padding leads to choose which kind of padding we want making yet one or even two new hyperparameters. 

I see two advantages of padding which are to deal with every patch the same way, even if they are on the border. The other one is that the idea behind overlap is that we are more confident about the predictions made inside the patch than on its border, with padding we would not keep any prediction on the border.

I think your code makes the first advantage irrelevant and the border effect on the volume border is probably minor so I'm good with your solution. Thanks a lot!

[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py]: Great! Do you think it's a good idea to add a `pad` kwarg to the `GridSampler` as well? The aggregator would read the attribute from the sampler and 1) not crop the patches and 2) crop the whole volume at the end.

[the remark this answers is not in the document]
Comment (GFabien) [reply torchio/data/inference/aggregator.py]: Yes I think the `pad` kwarg makes sense to be in the `GridSampler`. 

The way I see padding is that the sampler produces all the patches from the padded volume, the model makes the predictions and the aggregator crops the predictions to reconstruct the output. Cropping is made the same way for every patch using the `patch_overlap` so the aggregator just has to crop the final volume to get the final output.

[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/data/inference/aggregator.py]: I've pushed the changes, but the test is not passing. I'll debug this and then I think it'll be ready to go.

**Comment by fepegar** [comment]
@GFabien could you take a look? I think everything's working now.

**Comment by fepegar** [comment]
Many thanks again for your help, @GFabien!

**Pull request merged**